runtime/contract_update_validation.go

```diff
@@ -0,0 +1,434 @@
+	foundDictionaryType, ok := found.(*ast.DictionaryType)
+	if !ok {
+		return validator.getTypeMismatchError(expected, found)
+	}
+
+	err := expected.KeyType.Equal(foundDictionaryType.KeyType, validator)
+	if err != nil {
+		return err
+	}
+
+	return expected.ValueType.Equal(foundDictionaryType.ValueType, validator)
+}
+
+func (validator *ContractUpdateValidator) CheckRestrictedTypeEquality(expected *ast.RestrictedType, found ast.Type) error {
+	foundRestrictedType, ok := found.(*ast.RestrictedType)
+	if !ok {
+		return validator.getTypeMismatchError(expected, found)
+	}
+
+	err := expected.Type.Equal(foundRestrictedType.Type, validator)
+	if err != nil || len(expected.Restrictions) != len(foundRestrictedType.Restrictions) {
+		return validator.getTypeMismatchError(expected, found)
+	}
+
+	for index, expectedRestriction := range expected.Restrictions {
+		foundRestriction := foundRestrictedType.Restrictions[index]
+		err := expectedRestriction.Equal(foundRestriction, validator)
+		if err != nil {
+			return validator.getTypeMismatchError(expected, found)
+		}
+	}
+
+	return nil
+}
+
+func (validator *ContractUpdateValidator) CheckInstantiationTypeEquality(expected *ast.InstantiationType, found ast.Type) error {
+	foundInstType, ok := found.(*ast.InstantiationType)
+	if !ok {
+		return validator.getTypeMismatchError(expected, found)
+	}
+
+	err := expected.Type.Equal(foundInstType.Type, validator)
+	if err != nil || len(expected.TypeArguments) != len(foundInstType.TypeArguments) {
+		return validator.getTypeMismatchError(expected, found)
+	}
+
+	for index, typeArgs := range expected.TypeArguments {
+		otherTypeArgs := foundInstType.TypeArguments[index]
+		err := typeArgs.Type.Equal(otherTypeArgs.Type, validator)
+		if err != nil {
+			return validator.getTypeMismatchError(expected, found)
+		}
+	}
+
+	return nil
+}
+
+func (validator *ContractUpdateValidator) CheckFunctionTypeEquality(_ *ast.FunctionType, _ ast.Type) error {
+	// Non storable type
+	panic(errors.NewUnreachableError())
+}
+
+func (validator *ContractUpdateValidator) CheckReferenceTypeEquality(_ *ast.ReferenceType, _ ast.Type) error {
+	// Non storable type
+	panic(errors.NewUnreachableError())
+}
+
+func (validator *ContractUpdateValidator) loadCompositeDecls() {
+	if validator.oldNestedCompositeDecls == nil {
+		validator.oldNestedCompositeDecls = map[string]*ast.CompositeDeclaration{}
+		for _, nestedComposite := range getDeclarationMembers(validator.oldDeclaration).Composites() {
+			validator.oldNestedCompositeDecls[nestedComposite.Identifier.Identifier] = nestedComposite
+		}
+	}
+
+	if validator.newNestedCompositeDecls == nil {
+		validator.newNestedCompositeDecls = map[string]*ast.CompositeDeclaration{}
+		for _, nestedComposite := range getDeclarationMembers(validator.newDeclaration).Composites() {
+			validator.newNestedCompositeDecls[nestedComposite.Identifier.Identifier] = nestedComposite
+		}
+	}
+}
```

Comment:
Maybe replace this function with another index and getter function on `memberIndices`, just like it has an index for functions by identifier, e.g `CompositesByIdentifier()`